### DISCUSSION TITLE
usePrepareContractWrite status is idle then error, no error message

### DISCUSSION THREAD

**Discussion by lsbyerley** [wevm/wagmi#787] [Q&A] [answered]
Hello.. trying to figure out why the `usePrepareContractWrite` hook is failing. There is no error message and the `onError` never fires but the status changes from `idle` to `error`. Here is a brief code snippet. Any help is appreciated.

The write (aka. deposit) function is always undefined. I am using a dynamic address based on the chainId, or setting it to the constant value if the chain is not supported. I am setting enabled to false because I am firing it from a click handler. Setting the `addressOrName` parameter to a constant address var yields the same results.

Reproducible codesandbox: https://codesandbox.io/s/usepreparecontractwrite-null-example-xp084y

```js
const contractAddressConfig = chainId => {
	const config = {
		4: {
			name: 'rinkeby',
			address: RINKEBY_CONTRACT,
		},
		5: {
			name: 'goerli',
			address: GOERLI_CONTRACT,
		},
		80001: {
			name: 'polygon-mumbai',
			address: POLYGON_MUMBAI_CONTRACT,
		},
		42161: {
			name: 'Arbitrum',
			address: ARBITRUM_ONE_CONTRACT,
		},
	};
	return config[chainId]?.address || false;
};
```

```js
const contractConfig = {
  addressOrName: contractAddressConfig(chain?.id) || constants.AddressZero,
  contractInterface: ContractJSON.abi,
};

const { config: contractWriteConfig, status } = usePrepareContractWrite({
  ...contractConfig,
  functionName: 'deposit',
  enabled: false,
  onError: error => {
    // never gets here
    console.log('Error prepare', error);
  },
  onSettled(data, error) {
    // never gets here
    console.log('Settled', { data, error });
  },
});

const {
  data: depositData,
  write: deposit,
  isLoading: depositLoading,
  isSuccess: depositSuccess,
  error: depositError,
} = useContractWrite(contractWriteConfig);

console.log('LOG: prepare', status, deposit, contractConfig);
```

```
LOG: prepare idle undefined {addressOrName: '0xCorrectAddressIsHere', contractInterface: Array(10)}
LOG: prepare error undefined {addressOrName: '0xCorrectAddressIsHere', contractInterface: Array(10)}
```

**Comment by jxom** [comment]
Do you think you would be able to create a codesandbox reproducing this issue? Will be much easier to look into what's happening!

**Reply by lsbyerley** [reply]
> Do you think you would be able to create a codesandbox reproducing this issue? Will be much easier to look into what's happening!

Sure i can try. Just wanted to get the code snippet posted in case anything could be spotted :)

**Reply by lsbyerley** [reply]
Is there a codesandbox template out there setup with nextjs and wagmi to start from?

**Reply by tmm** [reply]
Here's a template: https://codesandbox.io/s/wagmi-ugtcvf

**Reply by lsbyerley** [reply]
Should i file a bug for this? @tmm

**Reply by tmm** [reply]
If you are able to create a minimal reproducible example, go for it.

**Reply by lsbyerley** [reply]
@tmm my bad, i thought i posted my codesandbox. 

here it is https://codesandbox.io/s/usepreparecontractwrite-null-example-xp084y

**Reply by davigp** [reply]
I'm facing the exact same issue, btw when the user rejects a transaction I can see the error in the console:
```
inpage.js:1 MetaMask - RPC Error: MetaMask Tx Signature: User denied transaction signature. {code: 4001, message: 'MetaMask Tx Signature: User denied transaction signature.'}
...
index.js:1 UserRejectedRequestError: User rejected request
    at sendTransaction (wagmi-core.esm.js:1040:1)
    at async writeContract (wagmi-core.esm.js:1130:1)
```
but not been able to capture the error using onError, onSettle, and error from hooks

**Reply by lsbyerley** [reply]
@davigp When you say "facing the exact same issue".. are you referring to the write function being undefined? If you use `writeAsync` is it not undefined?

**Comment by davigp** [comment]
the only way I found to get the error it was using writeAsync and try/catch

**Reply by lsbyerley** [reply]
@davigp were you able to figure out a resolution for this? Even using `writeAsync` did not reveal any errors for me.

**Comment by lsbyerley** [comment]
codesandbox with the undefined write function: https://codesandbox.io/s/usepreparecontractwrite-null-example-xp084y

**Reply by holma91** [reply]
Did you figure this out? Everytime I try to use these hooks with "dynamic arguments" I get the problem that the write function is undefined...

**Reply by lsbyerley** [reply]
I have not figured this out. Even when i hardcode the contract address for `addressOrName` i still get undefined. My codesandbox example is still dynamic, but i tried switching it out for the constant variable and it did not fix the issue.

**Reply by jxom** [reply]
Hi. I have forked the example to make it work. Check here!

https://codesandbox.io/s/usepreparecontractwrite-null-example-forked-mgmzf0?file=/pages/index.js

Note: if you set `enabled: false` then it will never prepare, and `write` will expectedly return `undefined`. To avoid running the prepare phase too early, we can guard it with variables (in this example, we only want it to run when there is a defined contract address & `name` contract argument)

**Reply by lsbyerley** [reply]
@jxom thank you for taking a look! I'm still trying to get this working in my ux.

I guess the enabled prop is a bit misleading. One would expect setting `enabled: false` would allow you to use the `write` function programmatically and pass args to it. Something like

```
const onSubmit = () => {
  write(argOne, { value: ethers.utils.parseEther(amount) });
}
```

Unless I'm not understanding this correctly, It seems you need to set the values of `args` and `overrides` in `usePrepareContractWrite` on init, before calling the `write` function? And once you set enabled to true it runs the `write` function?

**Reply by jxom** [reply]
It is important to understand the differences between `usePrepareContractWrite` and `useContractWrite`. 
- The `usePrepareContractWrite` is a [Prepare Hook](https://wagmi.sh/docs/prepare-hooks/intro), that only gives you the configuration required to run `useContractWrite` – it does not perform any transaction. Note: one of the main functions of `usePrepareContractWrite` is to estimate the gas required to execute the transaction.
- The `useContractWrite` hook is the one that performs the transaction (contract write).

If `enabled` is truthy on `usePrepareContractWrite`, this means that the "query" (prepare function) will run automatically, and return the configuration required for the write phase (`useContractWrite`). 
If `enabled` is falsy on `usePrepareContractWrite`, this means that the prepare function will not run automatically, hence, it will not return configuration required for the write phase (`useContractWrite`).

The `write` function on `useContractWrite` becomes defined when the prepare phase is successful (`usePrepareContractWrite` is successful).

It would be nice if we could dynamically pass in arguments to `write`, however, this leads to [UX pitfalls as we need to asynchronously estimate the gas limit and other values](https://wagmi.sh/docs/prepare-hooks/intro#ux-pitfalls-without-prepare-hooks).

**Reply by lsbyerley** [reply]
@jxom thank you for the explainer. Everything you say makes sense, I just have a hard time getting the functions to work as you say they do. Even setting a name field in your example still results in a deposit undefined error

<img width="632" alt="Screen Shot 2022-08-22 at 10 36 19 AM" src="https://user-images.githubusercontent.com/3066258/185947921-515ad0f9-681e-4cd2-bb86-99697a51ca51.png">

**Comment by lsbyerley** [comment, marked as answer]
The write function needs to be stored and referenced in a useRef.

```js
const depositFuncRef = useRef();

const {
  data: depositData,
  writeAsync: deposit,
  isLoading: depositLoading,
  isSuccess: depositSuccess,
  error: depositError,
} = useContractWrite(contractWriteConfig);

useEffect(() => {
  if (deposit && typeof deposit === 'function') {
    depositFuncRef.current = deposit;
  }
}, [deposit]);

const depositGratuity = async () => {
  try {
    await depositFuncRef.current();
   // catch the success in the `useContractEvent` hook listener and reset ref
  } catch (error) {
    alert('Transaction failed or denied!');
    // reset ref here also
  }
};
```

On success or fail, the ref should be set back to null and enabled again when ready to call it again
